staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -134,6 +142,25 @@ type crdHandler struct {
 	// The limit on the request size that would be accepted and decoded in a write request
 	// 0 means no limit.
 	maxRequestBodyBytes int64
+
+	apiserverID          string
+	storageVersionClient storageversion.Client
+
+	storageVersionUpdateQueues map[types.UID]*storageVersionUpdateQueue
+}
+
+type storageVersionUpdateRequest struct {
+	crd *apiextensionsv1.CustomResourceDefinition
+	// if non-nil, wait for the channel to be closed before processing the request.
+	wait <-chan struct{}
+	// if non-nil, close either channel to confirm whether the request processed or abandoned.
+	processed chan<- struct{}
+	abandoned chan<- struct{}
+}
+
+type storageVersionUpdateQueue struct {
+	cancel context.CancelFunc
```

Comment:
would expect this on a request, not on a queue. What does cancelling mean here?

Reply:
all the requests in one queue sharing the same context. Cancelling means aborting all the pending requests in the queue. 

It's used when a CRD is deleted: we close the queue for that CRD, to make sure the associated goroutine stops; we also don't know if the CRD will be re-created, bring up a new queue + goroutine, and race new requests with the old requests-- therefore we abort all the pending requests to avoid such race.